File: LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.cpp
/*
 * time: O(n), space: O(1)
 */
#include <iostream>
#include <vector>
#include <map>
using namespace std;

class Solution{
public:
    int lengthOfLongestSubstring(string s){
        bool hasFound[256] = {false};
        int maxLen = 0;
        int start = 0;
        int end = 0;
        while(end < s.size()){
            if(!hasFound[s[end]]){
                hasFound[s[end]] = true;
                end++;
                continue;
            }

            maxLen = max(maxLen, end - start);
            while(s[start] != s[end]){
                hasFound[s[start]] = false;
                start++;
            }
            start++;
            end++;
        }
        maxLen = max(maxLen, end - start);
        return maxLen;
    }
};

int main(){
    string s = "abcdbxyz";
    Solution solution;
    cout << s << " " << solution.lengthOfLongestSubstring(s) << endl;
    return 0;
}

File: algorithms/LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        from collections import defaultdict

        d = defaultdict(int)
        count = 0
        start, end = 0, 0
        max_len = 0

        while end < len(s):
            c1 = s[end]
            d[c1] += 1
            if d[c1] > 1:
                count += 1
            end += 1

            while count > 0:
                c2 = s[start]
                if d[c2] > 1:
                    count -= 1
                d[c2] -= 1
                start += 1

            max_len = max(max_len, end - start)

        return max_len
